tools/generator/Java.Interop.Tools.Generator.Transformation/MethodVisibilityFixup.cs

```diff
@@ -0,0 +1,28 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MonoDroid.Generation
+{
+	// Java allows a public member to override a protected member, but C# does not.
+	// This Fixup makes the override member protected as well.
+	static class MethodVisibilityFixup
+	{
+		public static void Fixup (GenBase gen)
+		{
+			foreach (var method in gen.GetAllMethods ()) {
+				var base_method = method.GetBaseMethodDeclaration ();
+
+				if (base_method == null)
+					continue;
+
+				if (method.Visibility == "public" && (base_method.Visibility == "protected" || base_method.Visibility == "protected internal")) {
+					Report.Warning (0, Report.WarningInconsistentAccessbility, $"Setting {gen.FullName}.{method.Name} visibility to {base_method.Visibility} to match base method");
+					method.Visibility = base_method.Visibility;
```

Comment:
This is where it would be "Interesting" to use the resulting `generator.exe` to generate `Mono.Android.dll`.  I suspect that this will result in huge API breaks.

For example, consider [`java.lang.Object.clone()`](https://developer.android.com/reference/java/lang/Object.html#clone()):

```java
/* partial */ class Object {
    protected java.lang.Object clone() throws java.lang.CloneNotSupportedException { ... }
}
```

It is *frequently* overridden in Android and turned into a `public` method, e.g. [`android.content.Intent.clone()`](https://developer.android.com/reference/android/content/Intent.html#clone()):

```java
/* partial */ class Intent extends Object {
    public Object clone() {...}
}
```

I'm fairly sure there are others; this is just the most obvious one, and there are *93* such `clone()` overloads that API-28 binds:

```
$ grep -r '\<n_Clone\> (' xamarin-android/src/Mono.Android/obj/Debug/android-28 | wc -l
      94
```

The intent, I'm sure, is to remove the CS0108 warnings that Csc emits for our Clone method:

```csharp
namespace Java.Lang {
    partial class Object {
        protected virtual unsafe Java.Lang.Object Clone () {/* ... */}
    }
}

namespace Android.Content {
    partial class Intent {
        public virtual unsafe Java.Lang.Object Clone () { /* ... */ }
    }
}
```

However, in this case what we need to do is emit `new` on the `Clone()` declaration, *not* change `Clone()` to be a `protected` method:

```csharp
namespace Android.Content {
    partial class Intent {
        public new virtual unsafe Java.Lang.Object Clone () { /* ... */ }
    }
}
```

Reply:
This does not change anything in regards to `Mono.Android.dll`.  It fixes this case:

```
    namespace Java.Lang
    {
        partial class Object
        {
            protected virtual Java.Lang.Object Clone() { return null; }
        }
    }

    namespace Android.Content
    {
        partial class Intent : Java.Lang.Object
        {
            public override Java.Lang.Object Clone() { return null; }
        }
    }
```

(That is, it is explicitly using `override`.)

This results in a compile error: 
```
CS0507	'Intent.Clone()': cannot change access modifiers when overriding 'protected' inherited member 'Object.Clone()'	
```

Because this case is already a compiler error it shouldn't break anything that was working.  Just a case of "we shouldn't be generating something by default that doesn't compile".  I hadn't thought of just declaring it as a `new` method.  I wonder if that's a better way to handle this?  Or does that break this case because the user explicitly wants to `override` instead of `new`?